substrate/primitives/runtime/src/generic/block.rs

```diff
@@ -77,6 +78,73 @@ impl<Block: BlockT> fmt::Display for BlockId<Block> {
 	}
 }
 
+/// Abstraction over a substrate block that allows us to lazily decode its extrinsics.
+#[derive(RuntimeDebug, Encode, Decode, scale_info::TypeInfo)]
+pub struct LazyBlock<Header, Extrinsic> {
+	/// The block header.
+	pub header: Header,
+	/// The accompanying extrinsics.
+	pub extrinsics: Vec<OpaqueExtrinsic>,
+
+	_phantom: PhantomData<Extrinsic>,
+}
+
+impl<Header, Extrinsic: Into<OpaqueExtrinsic>> LazyBlock<Header, Extrinsic> {
+	/// Creates a new instance of `LazyBlock` from its parts.
+	pub fn new(header: Header, extrinsics: Vec<Extrinsic>) -> Self {
+		Self {
+			header,
+			extrinsics: extrinsics.into_iter().map(|xt| xt.into()).collect(),
+			_phantom: Default::default(),
+		}
+	}
+}
+
+impl<Header, Extrinsic: Into<OpaqueExtrinsic>> From<Block<Header, Extrinsic>>
+	for LazyBlock<Header, Extrinsic>
+{
+	fn from(block: Block<Header, Extrinsic>) -> Self {
+		LazyBlock::new(block.header, block.extrinsics)
+	}
+}
+
+impl<Header, Extrinsic> EncodeLike<LazyBlock<Header, Extrinsic>> for Block<Header, Extrinsic>
+where
+	Block<Header, Extrinsic>: Encode,
+	LazyBlock<Header, Extrinsic>: Encode,
+{
+}
+
+impl<Header, Extrinsic> EncodeLike<Block<Header, Extrinsic>> for LazyBlock<Header, Extrinsic>
+where
+	Block<Header, Extrinsic>: Encode,
+	LazyBlock<Header, Extrinsic>: Encode,
+{
+}
+
+impl<Header, Extrinsic> traits::LazyBlock for LazyBlock<Header, Extrinsic>
+where
+	Header: HeaderT,
+	Extrinsic: core::fmt::Debug + LazyExtrinsic,
+{
+	type Extrinsic = Extrinsic;
+	type Header = Header;
+
+	fn header(&self) -> &Self::Header {
+		&self.header
+	}
+
+	fn header_mut(&mut self) -> &mut Self::Header {
+		&mut self.header
+	}
+
+	fn extrinsics(&self) -> impl Iterator<Item = Result<Self::Extrinsic, codec::Error>> {
+		self.extrinsics
+			.iter()
+			.map(|xt| Self::Extrinsic::decode_with_len(&xt.0, xt.0.len()))
```

Comment:
Instead of having this method `LazyExtrinsic::decode_with_len` we could just use the implementation of `Decode` here and for the input we would use a custom input that prefixes the length before decoding.

Or maybe we just don't use `OpaqueExtrinsic` but a type that keeps the full encoded value instead of removing the length prefix.

This way we avoid 2 decode implementation.

But anyway I think it is ok.